File: module-4/mcp-dispatchkit-core/src/tools/list-vehicles.ts
import { z } from "zod";
import { vehicles } from "../data/vehicles.js";

export const listVehiclesInputShape = {
  status: z
    .enum(["active", "idle", "maintenance"])
    .optional()
    .describe("If provided, only return vehicles in this status."),
};

const listVehiclesInput = z.object(listVehiclesInputShape);

export const listVehiclesDescription =
  "List vehicles in the DispatchKit fleet. Returns each vehicle's id, registration, status, capacityKg, current location (lat, lon, name), lastSeenAt timestamp, and currentRouteId (null if unassigned). Use the optional `status` filter to narrow results.";

export async function listVehiclesHandler(
  args: z.infer<typeof listVehiclesInput>
) {
  const filtered = args.status
    ? vehicles.filter((v) => v.status === args.status)
    : vehicles;

  return {
    content: [
      {
        type: "text" as const,
        text: JSON.stringify(filtered, null, 2),
      },
    ],
  };
}

File: module-4/python/mcp-dispatchkit-core/src/tools/list_vehicles.py
import json
from dataclasses import asdict
from typing import Optional

from ..data.vehicles import VEHICLES, VehicleStatus


def list_vehicles(status: Optional[VehicleStatus] = None) -> str:
    """List vehicles in the DispatchKit fleet.

    Returns each vehicle's id, registration, status, capacityKg, current
    location (lat, lon, name), lastSeenAt timestamp, and currentRouteId
    (null if unassigned). Use the optional `status` filter to narrow
    results.
    """
    filtered = VEHICLES if status is None else [v for v in VEHICLES if v.status == status]
    return json.dumps([asdict(v) for v in filtered], indent=2)
